slatedb/src/compactor.rs

```diff
@@ -747,10 +800,16 @@ impl CompactorEventHandler {
         }
 
         self.state.add_compaction(compaction.clone())?;
+        self.write_compactions_safely().await?;
         // Compactions and jobs are 1:1 right now.
         let job_id = compaction.id();
         tracing::Span::current().record("id", tracing::field::display(&job_id));
-        self.start_compaction(job_id, compaction).await?;
+        if let Err(err) = self.start_compaction(job_id, compaction).await {
+            self.state.remove_compaction(&job_id);
+            // Best effort to persist the removal; prefer fencing errors for visibility.
```

Comment:
what's the failure situation here? the "bad" compaction gets persisted (`write_compactions_safely`) then this one gets fenced so we never remove it. the new compactor reads the compactor state and sees it should execute the compaction, but it fails because it was a bad compaction for whatever reason. we should make sure that compactors just remove "bad" compaction jobs when executing in steady state (not just the compactor that first submitted the compaction)

Reply:
Yeah, I understood the comment and the scenario. What I'm saying is that the scenario could act as a poison pill if we're not careful. There are a few situations when we remove entries from `.compactions` (namely whenever we startup). If we didn't remove them on startup and instead tried resuming them, a compaction job that can't start would cause every new compactor to fail. I was just saying we should be careful about that.

Reply:
Ah, gotcha yes. Agreed. Not an issue for now, since we don't resume. I did consider doing something more like `CompactionEventHandler::reconcile_compations`, but found that a) `remove_compaction` doesn't return a result, so it's safe (modulo panics), and b) `write_compactions_safely` only returns `TransactionalObjectError`.

(b) could theoretically cause an infinite loop if we hit some oddity with the object store, but I think it's not possible (at first glance).

tl;dr I think it's safe. When we add resume, we'll have to make sure we track attempts and give up after a while. This is in the RFC.